src/main/java/org/kiwiproject/beta/slf4j/TimestampingLogger.java

```diff
@@ -0,0 +1,147 @@
+ * <p>
+ * This is intended to be used for a single logical operation which contains multiple steps, or for the same operation
+ * repeated over a collection. For example, an order operation with separate steps to find a user, create a new order
+ * linked to the user, and insert the order to a database. Or, some kind of data migration that loops over records in
+ * a source database table and writes to a new target database table.
+ * <p>
+ * <em>Currently, this is intended only to be used within a single thread.</em>
+ */
+@NotThreadSafe
+@Beta
+public class TimestampingLogger {
+
+    @SuppressWarnings("NonConstantLogger")
+    private final Logger logger;
+    private long previousTimestamp;
+
+    /**
+     * Create a new instance using the given {@link Logger}.
+     */
+    public TimestampingLogger(Logger logger) {
+        this.logger = requireNotNull(logger);
+    }
+
+    /**
+     * Logs a message and elapsed time message at TRACE level.
+     *
+     * @param message the message or message template
+     * @param args    the arguments to the message template, if any
+     * @see #logElapsed(Level, String, Object...)
+     */
+    public void traceLogElapsed(String message, Object... args) {
+        logElapsed(Level.TRACE, message, args);
+    }
+
+    /**
+     * Logs a message and elapsed time message at DEBUG level.
+     *
+     * @param message the message or message template
+     * @param args    the arguments to the message template, if any
+     * @see #logElapsed(Level, String, Object...)
+     */
+    public void debugLogElapsed(String message, Object... args) {
+        logElapsed(Level.DEBUG, message, args);
+    }
+
+    /**
+     * Logs the given message, and then logs the elapsed time since the previous log.
+     * This results in two separate log messages.
+     *
+     * @param level   the level at which to log the message and elapsed time message
+     * @param message the message or message template
+     * @param args    the arguments to the message template, if any
+     */
+    public void logElapsed(Level level, String message, Object... args) {
+        if (KiwiSlf4j.isEnabled(logger, level)) {
+            var now = System.nanoTime();
+            KiwiSlf4j.log(logger, level, message, args);
+            logElapsedSincePreviousTimestamp(logger, level, now, previousTimestamp);
+            previousTimestamp = now;
+        }
+    }
+
+    private static void logElapsedSincePreviousTimestamp(Logger logger,
+                                                         Level level,
+                                                         long now,
+                                                         long previousTimestamp) {
+        if (previousTimestamp > 0) {
+            var diffInNanos = now - previousTimestamp;
+            KiwiSlf4j.log(logger, level,
+                    " -- Time spent (since previous log): {} nanoseconds / {} millis",
```

Comment:
prefer a shorter appended string `[delta {} ns / {} ms]`

Reply:
OK, but think the "delta" might be a tad short. I can change it to `[since previous log: N nanos / M millis]`?

Reply:
or maybe just `[since previous: N nanos / M millis]

Reply:
Also, would you omit the brackets in the "separate log" methods?

Reply:
i think the separate message can be full length: `elapsed time since previous: ...`

Reply:
i.e.

2-line:
```
The original log message
[since previous: N nanos / M millis]
```

and:

single-line
```
The original log message [since previous: N nanos / M millis]
```

Reply:
ok...
1) always brackets
2) `[elapsed time since previous: N nanos / M millis]` for the 2 message vesion

Reply:
ok